service/orchestrator/certificates.go

```diff
@@ -0,0 +1,166 @@
+	"clouditor.io/clouditor/internal/logging"
+	"clouditor.io/clouditor/persistence"
+	"clouditor.io/clouditor/service"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+// ErrCertificationNotFound indicates the certification was not found
+var ErrCertificationNotFound = status.Error(codes.NotFound, "certificate not found")
+
+// CreateCertificate implements method for creating a new certificate
+func (svc *Service) CreateCertificate(ctx context.Context, req *orchestrator.CreateCertificateRequest) (
+	res *orchestrator.Certificate, err error) {
+
+	// Validate request
+	if err = service.ValidateRequest(req); err != nil {
+		return
+	}
+
+	// Check if client is allowed to access the corresponding cloud service (targeted in the certificate)
+	if !svc.authz.CheckAccess(ctx, service.AccessCreate, req) {
+		err = service.ErrPermissionDenied
+		return
+	}
+
+	// Persist the new certificate in our database
+	err = svc.storage.Create(req.Certificate)
+	if err != nil {
+		err = status.Errorf(codes.Internal, "could not add certificate to the database: %v", err)
+		return
+	}
+
+	logging.LogRequest(log, logrus.DebugLevel, logging.Create, req)
+
+	// Return certificate
+	res = req.Certificate
+	return
+}
+
+// GetCertificate implements method for getting a certificate, e.g. to show its state in the UI
+func (svc *Service) GetCertificate(ctx context.Context, req *orchestrator.GetCertificateRequest) (
+	res *orchestrator.Certificate, err error) {
+
+	// Validate request
+	if err = service.ValidateRequest(req); err != nil {
+		return
+	}
+
+	res = new(orchestrator.Certificate)
+	err = svc.storage.Get(res, "Id = ?", req.CertificateId)
+	if errors.Is(err, persistence.ErrRecordNotFound) {
+		return nil, ErrCertificationNotFound
+	} else if err != nil {
+		return nil, status.Errorf(codes.Internal, "database error: %v", err)
+	}
+
+	// Check if client is allowed to access the corresponding cloud service (targeted in the certificate)
+	areAllAllowed, allowedCloudServices := svc.authz.AllowedCloudServices(ctx)
+	if !areAllAllowed && !slices.Contains(allowedCloudServices, res.CloudServiceId) {
+		// Important to nil the response since it is set already
+		return nil, status.Error(codes.PermissionDenied, service.ErrPermissionDenied.Error())
+	}
+
+	return
+}
+
+// ListCertificates implements method for getting a certificate, e.g. to show its state in the UI
+func (svc *Service) ListCertificates(ctx context.Context, req *orchestrator.ListCertificatesRequest) (
+	res *orchestrator.ListCertificatesResponse, err error) {
+
+	// Validate request
+	if err = service.ValidateRequest(req); err != nil {
+		return nil, err
+	}
+
+	// We only list certificates the user is authorized to see (w.r.t. the cloud service)
+	var conds []any
+	areAllAllowed, cloudServices := svc.authz.AllowedCloudServices(ctx)
+	if !areAllAllowed {
```

Comment:
Here is the PermissionDenied check missing.

Reply:
We have to agree on how to do these things. These are the two options I see:
1) Return PermissionDeniedError: This means we would first have to retrieve the certificate, then look at its cloud service ID. If we are allowed, we then call the remove function. Otherwise we return PermissionDeniedError
2) Return NotFoundError: We build it into the DB query (`cloud_service_id IN ?`) which means we get a NotFoundError if the user is not auhorized. It could also means that the certificate is just not there.

@oxisto

Reply:
I'm for the first option. Otherwise, it is not possible to know why the certificate isn't removed. I think the same applies to the ListCertificates method.

Reply:
Well from a security (and privacy) point of view it should be indistinguishable whether the certificate is not there or whether you don’t have access, otherwise you are leaking the existence of a certificate with a certain ID.

@immqu I think we even have a check for this in the privacy property graph, correct?

Reply:
Maybe our access check is not that good. I think the ListCertificates endpoint shouldn't leak any information about certificates. There should be only two cases:
1. You have access and you get the List of certificates, maybe it is empty.
2. You do not have permission and do not receive information about any certificates or whether they exist.

I just worry that we will not recognize permission denied problems, e.g. if there is no cloud service id in the context. Suggestion: We check if areAllAllowed/all == false and the cloudServices/allowed list == empty. If yes, we return a permissionDeniedError, if not we return the list of (empty) certificates.

Reply:
> Well from a security (and privacy) point of view it should be indistinguishable ...
I agree to that

I think in the case of ListXXX methods it is pretty clear: You just get the list of things what you are allowed to see, i.e. in the case of ListCertificate: The list of certificates which are belonging to the cloud services you are allowed to see. Can be empty (there is no certificate yet that you are authorized to see w.r.t. cloud service) , can be all (e.g. admin flag), can be subset (e.g. you are only allowed to see certificates for cloud service 1 and 2 but not the rest)

Reply:
I added an explanation to the ListCertificates method.